### ISSUE TITLE
Quantization on multi-node GPUs

### ISSUE THREAD

**Issue by Devy99** [Vahe1994/AQLM#112]
Hello,
thank you for providing and well documenting the code for quantizing and fine-tuning the models.

I would ask whether you managed (and if it is possible) to quantize AQLM models on a multi-node hardware, thus splitting the computation cost and, eventually, the required memory over multiple GPUs distributed on different machines. 
An example of how to run your script with this configuration would be very appreciated.

Thanks!

**Comment by justheuristic** [comment]
Hi!

TL;DR alas, we do not support multi-node calibration. We support multi-gpu nodes, but not the other way around.

This is because our calibration is implemented with threading, not multiprocessing.  We currently have no bandwidth for that, but if anyone intends to implement it, we'd gladly accept a pull request. However, we expect that this will need  a lot of manual use of torch.distributed to work efficiently.
For instance, we rely on threading so that we can reuse shared memory, but that won't work multi-node, so it will need a different workaround. In other words, this may be tricky to get right.

If your model does not fit into a single node, the simplest workaround is to modify [model loading](https://github.com/Vahe1994/AQLM/blob/main/main.py#L882-L888) so that it only loads one layer at a time. This is because main.py processes one transformer layer at a time, as can be seen [in this for loop](https://github.com/Vahe1994/AQLM/blob/main/main.py#L188). Thus, if you load the appropriate layer (e.g. from disk) at the beginning of that loop, and [deleting it once it after quantized layer has been saved](https://github.com/Vahe1994/AQLM/blob/main/main.py#L310-L316), you can quantize very large models in due time.

If time is what you're after, we'd recommend running AQLM on a single node with higher mse tolerances (i.e. much faster), then running better/longer PV fine-tuning , which does support multi-node HW https://github.com/Vahe1994/AQLM/tree/pv-tuning?tab=readme-ov-file#3-refining-quantized-model-with-pv-tuning

**Comment by Devy99** [comment]
Thanks for the suggestion. I'll proceed as suggested then! 😄
